src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -274,34 +281,36 @@ module Make (Inputs : Inputs_intf.S) = struct
           | _ ->
               None )
       in
-      let from_parent = lookup_parent ancestor not_found in
-      List.fold_map self_found_or_none ~init:from_parent
-        ~f:(fun from_parent (id, self_found) ->
-          match (self_found, from_parent) with
-          | None, r :: rest ->
-              (rest, r)
-          | Some acc_found_locally, _ ->
-              (from_parent, (id, acc_found_locally))
-          | _ ->
-              failwith "unexpected number of results from DB" )
-      |> snd
+      if List.is_empty not_found then
+        List.map ~f:(fun (a, x) -> (a, Option.value_exn x)) self_found_or_none
+      else
+        let from_parent = lookup_parent ancestor not_found in
```

Comment:
This code refactor is purely for clarity that call to parent happens only if some records were not found "locally" and underlying ledger needs to be queried.